src/datatype.c

```diff
@@ -1057,11 +1057,12 @@ JL_DLLEXPORT jl_value_t *jl_get_nth_field_checked(jl_value_t *v, size_t i)
 
 void set_nth_field(jl_datatype_t *st, void *v, size_t i, jl_value_t *rhs) JL_NOTSAFEPOINT
 {
+    if (rhs == NULL) // TODO: this should be invalid, but it happens frequently in ircode.c
```

Comment:
What's the purpose of this behavior change in the context of this change (don't really have an independent opinion on the change, but seems odd here).

Reply:
We're going to potentially attempt to read the type tag (for a write barrier or to figure out how to copy the data). I suppose we should actually store the NULL, but the use case (ircode) doesn't care. It's mostly here because I had added an assert here in April, and it started failing when I tried to rebase this.

Reply:
In the ircode usage is the previous field always NULL? If so maybe assert that here?

Reply:
Yeah, it's a new object (GC hasn't run yet since allocation), otherwise we'd sometimes segfault on the write barrier a couple lines later.

Reply:
Alright, I'd say let's add that assert then, since that way there's at least no silent behavior change if anybody else thought they'd be clever by using this to unset a field.

Reply:
Might as well. We should probably also look into avoiding it in that file also, but it's not causing problems yet.